### ISSUE TITLE
Restartable?

### ISSUE THREAD

**Issue by dposada** [twosigma/Cook#291]
Or, can the job tolerate being preempted, by Cook or host. This could be used to avoid scheduling non-restartable jobs on preemtable hosts, for example. The idea of having Cook not preempt the job needs more thought.

**Comment by dposada** [comment]
@wyegelwel @sradack @shamsimam @pschorf Should we be more specific with the new field and focus only on preemtable hosts? For example, `allow-preemtable-host` (boolean).

**Comment by shamsimam** [comment]
Are we going to have the `restartable?` flag too? In that case, doesn't `restartable?` `false` automatically mean `allow-preemptable-host` has to be `false`? I would like to avoid the scenario where we can end up having two fields with conflicting values.

**Comment by dposada** [comment]
I think if in the future we were to add `restartable` (i.e. by cook) we would have to also add validation to catch the conflict you describe. We already have the case of fields that can conflict with PR #295, where we are adding validation to ensure that `expected-runtime` <= `max-runtime`.

**Comment by dposada** [comment]
@wyegelwel thoughts on this?

**Comment by sradack** [comment]
I don't like `allow-preemptable-host`.  It would be better to gather general information about a job's _relative_ dislike for preemptions.

The name shouldn't imply that a job can never be preempted.  This is too hard to guarantee.

If I schedule a job A before I know about higher-priority job B, should I preempt A in order to schedule B?  Ideally the new field gives us better information to make the decision, in the single-user case.  For a given user, do we preempty lower priority jobs for higher priority jobs?  Should we?  I'd like to consider never doing this in favor of allowing the user to kill/reschedule their lower priority jobs. 

If I have a job C that I could attempt to schedule on a transient host, should I schedule it or not?  That seems to be another question at hand.  My question would be: when would we ever _not_ just schedule it?  Do we actually have insight into its future likelihood of preemption?  

If you answered that we would schedule preemption-hating jobs on more stable machines and we somehow have that information, it seems like we simply need a way to compare the relative preemption-hate among jobs to in order to arrive an an optimal distribution among hosts.  

What I think will be *hard* is trying, in isolation, to utilize the preemption-hate-factor to determine whether we should schedule something on a transient host.

**Comment by dposada** [comment]
In the case of AWS EC2 spot instances, we can get some insight into the likelihood of an instance being terminated -- from http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-interruptions.html:

> The following are the possible reasons that Amazon EC2 will terminate your Spot instances:
> Price—The Spot price is greater than your bid price.
> Capacity—If there are not enough unused EC2 instances to meet the demand for Spot instances, Amazon EC2 terminates Spot instances, starting with those instances with the lowest bid prices. If there are several Spot instances with the same bid price, the order in which the instances are terminated is determined at random.
> Constraints—If your request includes a constraint such as a launch group or an Availability Zone group, these Spot instances are terminated as a group when the constraint can no longer be met.

**Comment by sradack** [comment]
How does what you quoted translate into building a fn `predicted-termination-time` that tells me when a host might become unavailable?  Are we planning on building such a function?

**Comment by dposada** [comment]
I assume we're not going to build `predicted-termination-time`. 

I think the simplest example of how we would use "preemption aversion" is the example from the top of the issue -- simply avoid scheduling jobs which are averse to preemption on preemptable (e.g. spot) hosts. If in the future we had a way to determine likelihood of preemption, then we could do something more complex.

**Comment by sradack** [comment]
As @wyegelwel noted, you may actually want to schedule preemption-averse jobs on spot hosts.

One important question is whether or not preemption-aversion is modeled as a distinct value (binary) or continuously.

To through out another idea, what about a field called `preemption-aversion` that is scaled 0-100 (or on the same scale as priority)?  As a operational matter, we might choose to not schedule jobs above an arbitrary value like 80 on spot hosts, and possibly scale the threshold depending on `expected-runtime`.

**Comment by sradack** [comment]
Another idea for the possible values of `preemption-aversion`:

NO_SPOT (High aversion with ability to specifically exclude unstable instances.  Should we even allow this?)
HIGH_AVERSION (All work lost if preempted.)
LOW_AVERSION (Some work lost if preempted.)
INDIFFERENT (Presumably no work lost if preempted.  Is this different from LOW_AVERSION?)

--- 

On further thought, is there such a thing as tasks that don't lose all their work?  Do we have examples in the wild of tasks with different aversions independent of their `expected-runtime`?

**Comment by pschorf** [comment]
I don't think the `preemption-aversion` metric would make sense from a client perspective. What does it mean for me to be 40 averse to preemption instead of 80 averse aside from one being more averse from the other? Does it mean I want them to be treated differently or not? I think actually using the number would lead to some consternation from people if we moved the value and changed the way a job was treated, and if we can't change the value why not just make it boolean?

I think the root problem here is that we're trying to let people tell the scheduler how to do something, without letting them tell the scheduler how to do something. Maybe it's worth just making the decision based on expected runtime, and if there's a real problem for people being scheduled on spot instances evaluating how to support them, whether with something explicit like this or a more oblique metric like `preemption-aversion`.

**Comment by shamsimam** [comment]
I like `preemption-aversion` with LOW, MEDIUM, HIGH:

- LOW = Feel free to preempt the job, don't care where you schedule me. Usually the quickest job to get scheduled as it is least restrictive.
- MEDIUM = Prefer not to be preempted, but okay if you do. E.g. let schedule choose when to pre-empt me if I submit a higher priority job. Another e.g., schedule me on a spot host and okay if I need to restarted when the host goes away.
- HIGH = Do not pre-empt me even if a higher priority job shows up later, schedule me on guaranteed hosts and only when I am inside my quota limit. User accepts that this setting can cause delay in the job being scheduled.

**Comment by sradack** [comment]
I like the HML the most so far. Does it cover all known cases?

**Comment by dposada** [comment]
@shamsimam I'm not sure if I understand the difference between your LOW and MEDIUM - both seem to say to the scheduler, "you can preempt me if you think it's best to".

**Comment by pschorf** [comment]
My worry with High, Medium, Low is that if we have a new feature later, there is some chance that we will end up "splitting" a bucket (presumably medium) because some users in Medium want it and some users don't (so you have Medium and Low-Medium or whatever.) It seems to me that if people really do care about these settings, at some point someone will want a weird combination that doesn't fit in HML and we will either have to make new buckets or just make them pick the closest to what they want, whereas with specific feature flags you don't have that problem.

**Comment by shamsimam** [comment]
@dposada MEDIUM allows the scheduler to not preempt the job even if a higher priority job from the user arrives.

@pschorf we can name them with feature flags. They just need to carry the essence of levels as the field has a notion of degrees of aversion.

**Comment by dposada** [comment]
I think @pschorf is suggesting providing a specific field for each specific setting (e.g. `allow-higher-priority-preempt?`) rather than one field with levels.

**Comment by shamsimam** [comment]
My reservation with multiple fields is that not all combinations will be valid. Instead if we list out all the combinations it'll be much simpler to explain to users as well as implement/debug in the code.

**Comment by dposada** [comment]
I agree that it's possible to have invalid combinations, but that would be easy for us to guard against with validation. In the two concrete examples we have so far (`allow-preemtable-host` and `allow-higher-priority-preempt`), any combo is valid. As for explaining to users, I actually think that more specific fields are easier to explain than what combination of flags is meant by the level "medium", for example.

**Comment by wyegelwel** [comment]
I'm skeptical of both `allow-preemptable-host` and `allow-higher-priority-preemption`.

For `allow-higher-priority-preemption`, we are providing users with more complicated ways of managing their running and queues tasks through priority. Instead, I propose we bite the bullet and design queue management primitives that capture the intention of users instead of trying to bake it into priority. For example,  having primitives for what jobs are allowed to exceed the non-preemptable share. Another example is what percent of jobs to preempt in order that a set of interactive jobs can run. These two cases are not well supported (or possible) with the priority. We could introduce a bunch of flags and knobs to turn in order to support these but designing the right primitives seems a better way forward.

For `allow-preemptable-host`, this is effectively no different than having a constraint that says only run the job on hosts with attribute "preemptable?" == "true". There are a few reasons we should favor the constraint. 

1. `allow-preemptable-host` doesn't provide choice for Cook. It MUST place the job on a non-preemtable host. This is compared to something akin to `restartable` where the job is simply informing Cook it doesn't do well with being restarted and Cook has the freedom to use this information or not.
2. Cook doesn't know which hosts are preemptable currently. This means users can set `allow-preemptable-host` to false and Cook could happily schedule the job on a preemptable host.
3. If we have `allow-preemtable-host` we will need to add a required field for the operator of cook to inform the scheduler what attributes means the host is preemptable. Compared to something like `restartable` where not providing the information would simply mean cook couldn't take the information into account to make better decisions.

In summation, I propose we punt on this for the time being. Add a way to specify constraints on jobs (we only have constraints for groups at the moment) and encourage users to use that for specifying they don't want to be on a preemptable host. 

------
Below this point is musings of a mad man.

Perhaps we don't need a `restartable` field at all. The contract could simply be Cook will try to ensure a job is placed on a host that will be around for at least the jobs expected run time.  The problem with this is it may encourage users to further overestimate expected run time and it conflates two intentions.
